### PR TITLE
Bug 1840315 – added telemetry to the splash screen data fetching

### PR DESCRIPTION
**Pull request by mavduevskiy**
Follow up to [2615](https://github.com/mozilla-mobile/firefox-android/pull/2615).

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1840315

### CODE REVIEW

**Comment by mavduevskiy** [comment]
DATA REVIEW REQUEST
1. What questions will you answer with this data?

- This data will help us understand how many users get experiment and marketing data fetched during extended first launch splash screen.

2. Why does Mozilla need to answer these questions? Are there benefits for users?
   Do we need this information to address product or business requirements?

– If we expect the majority of users fetch the experiment/marketing data during an reasonable extended splash screen, it would allow us to run off-train experiments and customize user experience depending on the source of install.

3. What alternative methods did you consider to answer these questions?
   Why were they not sufficient?

- There are no alternatives, we do not have tools to answer that question differently.

4. Can current instrumentation answer these questions?

- Not fully. Nimbus has some data on the fetching time, but we don't have reliable data on the speed of google reference library.

5. List all proposed measurements and indicate the category of data collection for each
   measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement Name | Measurement Description | Data Collection Category | Tracking Bug
---------------- | ----------------------- | ------------------------ | ------------
`splash_screen.first_launch_extended` | The splash screen was shown for an extended period of time, providing more time to download marketing and experiment data.  | technical | https://bugzilla.mozilla.org/show_bug.cgi?id=1840315

6. Please provide a link to the documentation for this data collection which
   describes the ultimate data set in a public, complete, and accurate way.

This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).

7. How long will this data be collected?

This collection will be collected for the next 10 releases.
android-probes@mozilla.com, cgordon@mozilla.com will be responsible for the collection.

8. What populations will you measure?

Users that will be enrolled into `splash-screen` experiment.

9. If this data collection is default on, what is the opt-out mechanism for users?

These collections are Glean. The opt-out can be found in the product's preferences.

10. Please provide a general description of how you will analyze this data.

Glean and Amplitude

11. Where do you intend to share the results of your analysis?

Glean, Amplitude and with mobile teams.

12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
    are proposing to use for this data collection?

No.

**Comment by rocketsroger** [comment]
> DATA REVIEW REQUEST
> 
>     1. What questions will you answer with this data?
> 
> 
>     * This data will help us understand how many users get experiment and marketing data fetched during extended first launch splash screen.
> 
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users?
>        Do we need this information to address product or business requirements?
> 
> 
> – If we expect the majority of users fetch the experiment/marketing data during an reasonable extended splash screen, it would allow us to run off-train experiments and customize user experience depending on the source of install.
> 
>     3. What alternative methods did you consider to answer these questions?
>        Why were they not sufficient?
> 
> 
>     * There are no alternatives, we do not have tools to answer that question differently.
> 
> 
>     4. Can current instrumentation answer these questions?
> 
> 
>     * Not fully. Nimbus has some data on the fetching time, but we don't have reliable data on the speed of google reference library.
> 
> 
>     5. List all proposed measurements and indicate the category of data collection for each
>        measurement, using the Firefox data collection categories found on the Mozilla wiki.
> 
> 
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug
> `splash_screen.first_launch_extended` 	The splash screen was shown for an extended period of time, providing more time to download marketing and experiment data. 	technical 	https://bugzilla.mozilla.org/show_bug.cgi?id=1840315
> 
>     6. Please provide a link to the documentation for this data collection which
>        describes the ultimate data set in a public, complete, and accurate way.
> 
> 
> This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).
> 
>     7. How long will this data be collected?
> 
> 
> This collection will be collected for the next 10 releases. [android-probes@mozilla.com](mailto:android-probes@mozilla.com), [cgordon@mozilla.com](mailto:cgordon@mozilla.com) will be responsible for the collection.
> 
>     8. What populations will you measure?
> 
> 
> Users that will be enrolled into `splash-screen` experiment.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
> These collections are Glean. The opt-out can be found in the product's preferences.
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
> Glean and Amplitude
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
> Glean, Amplitude and with mobile teams.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
>         are proposing to use for this data collection?
> 
> 
> No.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/apps/fenix )

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, through the "Send Usage Data" preference in the application settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

cgordon@mozilla.com will be responsible for monitoring the data over time

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 1, Technical data

5) Is the data collection request for default-on or default-off?

default-on

6) Does the instrumentation include the addition of **any *new* identifiers**?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

## Result

data-review+

File: fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 1]
```diff
@@ -433,7 +434,13 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
             splashScreen.setKeepOnScreenCondition {
                 val dataFetched = components.settings.utmParamsKnown &&
                     components.settings.nimbusExperimentsFetched
-                !maxDurationReached && !dataFetched
+                val keepOnScreen = !maxDurationReached && !dataFetched
+                if (!keepOnScreen) {
+                    SplashScreen.firstLaunchExtended.record(
+                        SplashScreen.FirstLaunchExtendedExtra(dataFetched = dataFetched),
```
Comment (rocketsroger) [inline fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:440]: Should we be recording telemetry in a function that's passed to `setKeepOnScreenCondition`?  Are we sure that this is only going to be called once?

Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:440]: It's a condition, so it shouldn't get called once we return `false`. From the doc:
`The splash will stay visible until the condition isn't met anymore.`

Is there a nimbus side way to ensure an even is fired once?

Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:440]: I will take my guess that it's `lifetime: application` in the metrics file. Updated it

Comment (mavduevskiy) [reply fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt:440]: It looks like `lifetime: application` adds some additional requirements to the event. removed

**Comment by rocketsroger** [comment]
> DATA REVIEW REQUEST
> 
>     1. What questions will you answer with this data?
> 
> 
>     * This data will help us understand how many users get experiment and marketing data fetched during extended first launch splash screen.
> 
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users?
>        Do we need this information to address product or business requirements?
> 
> 
> – If we expect the majority of users fetch the experiment/marketing data during an reasonable extended splash screen, it would allow us to run off-train experiments and customize user experience depending on the source of install.
> 
>     3. What alternative methods did you consider to answer these questions?
>        Why were they not sufficient?
> 
> 
>     * There are no alternatives, we do not have tools to answer that question differently.
> 
> 
>     4. Can current instrumentation answer these questions?
> 
> 
>     * Not fully. Nimbus has some data on the fetching time, but we don't have reliable data on the speed of google reference library.
> 
> 
>     5. List all proposed measurements and indicate the category of data collection for each
>        measurement, using the Firefox data collection categories found on the Mozilla wiki.
> 
> 
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug
> `splash_screen.first_launch_extended` 	The splash screen was shown for an extended period of time, providing more time to download marketing and experiment data. 	technical 	https://bugzilla.mozilla.org/show_bug.cgi?id=1840315
> 
>     6. Please provide a link to the documentation for this data collection which
>        describes the ultimate data set in a public, complete, and accurate way.
> 
> 
> This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).
> 
>     7. How long will this data be collected?
> 
> 
> This collection will be collected for the next 10 releases. [android-probes@mozilla.com](mailto:android-probes@mozilla.com) will be responsible for the collection.
> 
>     8. What populations will you measure?
> 
> 
> Users that will be enrolled into `splash-screen` experiment.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
> These collections are Glean. The opt-out can be found in the product's preferences.
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
> Glean and Amplitude
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
> Glean, Amplitude and with mobile teams.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
>         are proposing to use for this data collection?
> 
> 
> No.



> DATA REVIEW REQUEST
> 
>     1. What questions will you answer with this data?
> 
> 
>     * This data will help us understand how many users get experiment and marketing data fetched during extended first launch splash screen.
> 
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users?
>        Do we need this information to address product or business requirements?
> 
> 
> – If we expect the majority of users fetch the experiment/marketing data during an reasonable extended splash screen, it would allow us to run off-train experiments and customize user experience depending on the source of install.
> 
>     3. What alternative methods did you consider to answer these questions?
>        Why were they not sufficient?
> 
> 
>     * There are no alternatives, we do not have tools to answer that question differently.
> 
> 
>     4. Can current instrumentation answer these questions?
> 
> 
>     * Not fully. Nimbus has some data on the fetching time, but we don't have reliable data on the speed of google reference library.
> 
> 
>     5. List all proposed measurements and indicate the category of data collection for each
>        measurement, using the Firefox data collection categories found on the Mozilla wiki.
> 
> 
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug
> `splash_screen.first_launch_extended` 	The splash screen was shown for an extended period of time, providing more time to download marketing and experiment data. 	technical 	https://bugzilla.mozilla.org/show_bug.cgi?id=1840315
> 
>     6. Please provide a link to the documentation for this data collection which
>        describes the ultimate data set in a public, complete, and accurate way.
> 
> 
> This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).
> 
>     7. How long will this data be collected?
> 
> 
> This collection will be collected for the next 10 releases. [android-probes@mozilla.com](mailto:android-probes@mozilla.com) will be responsible for the collection.
> 
>     8. What populations will you measure?
> 
> 
> Users that will be enrolled into `splash-screen` experiment.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
> These collections are Glean. The opt-out can be found in the product's preferences.
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
> Glean and Amplitude
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
> Glean, Amplitude and with mobile teams.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
>         are proposing to use for this data collection?
> 
> 
> No.

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? 

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/apps/fenix )

2) Is there a control mechanism that allows the user to turn the data collection on and off? 

Yes, through the "Send Usage Data" preference in the application settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

N/A, collection set to end or be renewed by version 126

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 1, Technical data

5) Is the data collection request for default-on or default-off?

default-on

6) Does the instrumentation include the addition of **any *new* identifiers**?

No

7) Is the data collection covered by the existing Firefox privacy notice? 

Yes

8) Does the data collection use a third-party collection tool? 

No

## Result

data-review+

**Review by rocketsroger — APPROVED** [review]
🚢

**Comment by rocketsroger** [comment]
@Mergifyio backport releases_v116

**Pull request merged**